src/storage/src/source/postgres/snapshot.rs

```diff
@@ -0,0 +1,376 @@
+//!    ┗━┯━━━━━━━━━━┯━┛   ╰─────────╯
+//!   raw│          │
+//!  COPY│          │
+//!  data│          │
+//! ╭────┴─────╮    │
+//! │distribute│    │
+//! ╰────┬─────╯    │
+//! ┏━━━━┷━━━━┓     │
+//! ┃  COPY   ┃     │
+//! ┃ decoder ┃     │
+//! ┗━━━━┯━━━━┛     │
+//!      │ snapshot │rewind
+//!      │ updates  │requests
+//!      v          v
+//! ```
+
+use std::any::Any;
+use std::collections::BTreeMap;
+use std::pin::pin;
+use std::rc::Rc;
+
+use anyhow::anyhow;
+use differential_dataflow::{AsCollection, Collection};
+use futures::TryStreamExt;
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::{Broadcast, ConnectLoop, Feedback};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::{Antichain, Timestamp};
+use tokio_postgres::types::PgLsn;
+use tokio_postgres::Client;
+use tracing::trace;
+
+use mz_expr::MirScalarExpr;
+use mz_ore::result::ResultExt;
+use mz_postgres_util::desc::PostgresTableDesc;
+use mz_repr::{Datum, DatumVec, Diff, Row};
+use mz_sql_parser::ast::{display::AstDisplay, Ident};
+use mz_storage_client::types::connections::ConnectionContext;
+use mz_storage_client::types::sources::{MzOffset, PostgresSourceConnection};
+use mz_timely_util::antichain::AntichainExt;
+use mz_timely_util::builder_async::{Event as AsyncEvent, OperatorBuilder as AsyncOperatorBuilder};
+use mz_timely_util::operator::StreamExt as TimelyStreamExt;
+
+use crate::source::postgres::replication::RewindRequest;
+use crate::source::postgres::DefiniteError;
+use crate::source::types::SourceReaderError;
+use crate::source::RawSourceCreationConfig;
+
+/// Renders the snapshot dataflow. See the module documentation for more information.
+pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
+    mut scope: G,
+    config: RawSourceCreationConfig,
+    connection: PostgresSourceConnection,
+    context: ConnectionContext,
+    resume_upper: Antichain<MzOffset>,
+    table_info: BTreeMap<u32, (usize, PostgresTableDesc, Vec<MirScalarExpr>)>,
+) -> (
+    Collection<G, (usize, Result<Row, SourceReaderError>), Diff>,
+    Stream<G, RewindRequest>,
+    Stream<G, Rc<anyhow::Error>>,
+    Rc<dyn Any>,
+) {
+    let op_name = format!("TableReader({})", config.id);
+    let mut builder = AsyncOperatorBuilder::new(op_name, scope.clone());
+
+    let (feedback_handle, snapshot_feedback) = scope.feedback(Default::default());
+
+    let mut snapshot_input = builder.new_input(&snapshot_feedback, Pipeline);
+    let (mut raw_handle, raw_data) = builder.new_output_connection(vec![Antichain::new()]);
+    let (mut rewinds_handle, rewinds) = builder.new_output_connection(vec![Antichain::new()]);
+
+    // The export id must be sent to all workes, so we broadcast the feedback connection
+    let (mut snapshot_handle, snapshot) = builder.new_output_connection(vec![Antichain::new()]);
+    snapshot.broadcast().connect_loop(feedback_handle);
+
+    let is_snapshot_leader = config.responsible_for("snapshot_leader");
```

Comment:
It seems odd to me that we choose the same random worker for all PG sources. The idea is just that we improve this over choosing `0`?

Reply:
We choose a different random worker for each pg source because the `responsible_for` function factors in the hash of the global id together with the partition

Reply:
Very clever, ty for the ponter